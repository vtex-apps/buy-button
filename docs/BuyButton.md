# Buy Button

## Description

`BuyButton` is a VTEX Component that is resposible to handle events of adding products in the minicart. This Component can be imported and used by any VTEX App.

:loudspeaker: **Disclaimer:** Don't fork this project, use, contribute, or open issue with your feature request.

## Table of Contents

- [Usage](#usage)
  - [Blocks API](#blocks-api)
    - [Configuration](#configuration)
  - [Styles API](#styles-api)
    - [CSS Namespaces](#css-namespaces)

## Usage

You should follow the usage instruction in the main [README](/README.md#usage).

Then, add `buy-button` block into your app theme, as we do in our [Product Details app](https://github.com/vtex-apps/product-details/blob/master/store/blocks.json).

### Blocks API

When implementing this component as a block, various inner blocks may be available. The following interface lists the available blocks within `BuyButton` and describes if they are required or optional.

```json
  "buy-button": {
    "component": "BuyButton"
  }
```

For now this block does not have any required or optional blocks.

#### Configuration

Through the Storefront, you can change the `BuyButton`'s behavior and interface. However, you also can make in your theme app, as Store theme does.

| Prop name            | Type      | Description                                                                      | Default value      |
| -------------------- | --------- | -------------------------------------------------------------------------------- | ------------------ |
| `isOneClickBuy`      | `Boolean` | Should redirect to the checkout page or not                                      | `false`              |
| `shouldOpenMinicart` | `Boolean` | Should open the Minicart after clicking the button                               | `false`              |
| `large`              | `Boolean` | Sets button to large style, filling whole width (like a `block`)                 | -                  |
| `available`          | `Boolean` | If component is available or not                                                 | `true`               |
| `showToast`          | `Boolean` | If toast with feedback should be shown after add item request is processed       | -                  |
| `showItemsPrice`     | `Boolean` | If you want to show the total price of items to be added to cart                 | `false`              |
| `customToastURL`     | `String`  | Set the link associated with the Toast created when adding an item to your cart. | `/checkout/#/cart` |
| `shouldAddToCart`    | `Boolean` | If it should add to cart when clicked                                             | true          |
| `showTooltipOnSkuNotSelected` | `Boolean` | If it should show a tooltip when you click the button but there's no SKU selected | `true` |

### Styles API

You should follow the Styles API instruction in the main [README](/README.md#styles-api).

#### CSS Namespaces

| Class Name            | Description                                   | Note                                            | Modifiers                  |
| --------------------- | --------------------------------------------- | ----------------------------------------------- | -------------------------- |
| `buyButtonText`       | Contains the text displayed inside buy button |                                                 |                            |
| `buttonDataContainer` | Wraps data inside the buy button              | Only appears when `showItemsPrice` prop is true |                            |
| `buttonItemsPrice`    | Wraps the price displayed inside buy button   | Only appears when `showItemsPrice` prop is true |                            |
