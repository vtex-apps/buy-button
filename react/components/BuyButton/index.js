import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ApolloConsumer } from 'react-apollo'
import find from 'lodash/find'
import Button from '@vtex/styleguide/lib/Button'
import { injectIntl, intlShape } from 'react-intl'

import ADD_TO_CART_MUTATION from './mutations/addToCartMutation.gql'
import ORDER_FORM_QUERY from './queries/orderFormQuery.gql'

const EVENT_SUCCESS = 'item:add'
const EVENT_ERROR = 'message:error'

/**
 * BuyButton Component. Adds a list of items to the cart.
 */
export class BuyButton extends Component {
  constructor(props) {
    super(props)
    this.state = { isLoading: false }
  }

  static defaultProps = {
    quantity: 1,
    seller: 1,
  }

  translateMessage = id => this.props.intl.formatMessage({ id: id })

  toastMessage = (success, err) => {
    const event = new Event(success ? EVENT_SUCCESS : EVENT_ERROR)
    event.detail = {
      success,
      message: success ? '' : this.translateMessage('buybutton.add-failure'),
      err,
    }
    document.dispatchEvent(event)

    this.setState({ isLoading: false })
  }

  handleAddToCart = client => {
    const { quantity, seller, skuId } = this.props

    const variables = {
      items: [
        {
          id: parseInt(skuId),
          index: 1,
          quantity,
          seller,
        },
      ],
    }

    this.setState({ isLoading: true })

    client
      .query({
        query: ORDER_FORM_QUERY,
      })
      .then(
        queryRes => {
          const {
            data: {
              orderForm: { orderFormId },
            },
          } = queryRes

          variables.orderFormId = orderFormId

          client
            .mutate({
              mutation: ADD_TO_CART_MUTATION,
              variables,
            })
            .then(
              mutationRes => {
                const { items } = mutationRes.data.addItem
                this.toastMessage(find(items, { id: skuId }))
                this.setState({ isLoading: false })
              },
              mutationErr => {
                this.toastMessage(false, mutationErr)
                this.setState({ isLoading: false })
              }
            )
        },
        queryErr => {
          this.toastMessage(false, queryErr)
        }
      )
  }

  render() {
    const { isLoading } = this.state

    return (
      <ApolloConsumer>
        {client => (
          <div>
            {isLoading ? (
              <Button disabled isLoading={isLoading}>
                {this.props.children}
              </Button>
            ) : (
              <Button primary onClick={() => this.handleAddToCart(client)}>
                {this.props.children}
              </Button>
            )}
          </div>
        )}
      </ApolloConsumer>
    )
  }
}

BuyButton.propTypes = {
  /** Message that will be displayed inside of the button **/
  children: PropTypes.PropTypes.node.isRequired,
  /** The quantity of products to be added to the cart */
  quantity: PropTypes.number,
  /** The specification of which product will be added to the cart */
  skuId: PropTypes.string.isRequired,
  /** Which seller is being referenced by the button */
  seller: PropTypes.number,
  /* Internationalization */
  intl: intlShape.isRequired,
}

export default injectIntl(BuyButton)
