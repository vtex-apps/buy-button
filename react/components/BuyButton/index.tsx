import React, { useContext, useCallback, useState, useEffect } from 'react'
import { FormattedMessage } from 'react-intl'
import { path } from 'ramda'
// @ts-expect-error ts-migrate(7016) FIXME: Try `npm install @types/react-content-loader` if i... Remove this comment to see the full error message
import ContentLoader from 'react-content-loader'
import { useRuntime } from 'vtex.render-runtime'
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'vtex.store-resources/PWAContex... Remove this comment to see the full error message
import { usePWA } from 'vtex.store-resources/PWAContext'
import { useCssHandles } from 'vtex.css-handles'
import useProduct from 'vtex.product-context/useProduct'
import { useProductDispatch } from 'vtex.product-context/ProductDispatchContext'
// @ts-expect-error ts-migrate(2305) FIXME: Module '"vtex.styleguide"' has no exported member ... Remove this comment to see the full error message
import { Button, ToastContext, Tooltip } from 'vtex.styleguide'

import useMarketingSessionParams from './hooks/useMarketingSessionParams'

const CONSTANTS = {
  SUCCESS_MESSAGE_ID: 'store/buybutton.buy-success',
  SELECT_SKU_ERROR_ID: 'store/buybutton.select-sku-variations',
  OFFLINE_BUY_MESSAGE_ID: 'store/buybutton.buy-offline-success',
  DUPLICATE_CART_ITEM_ID: 'store/buybutton.buy-success-duplicate',
  ERROR_MESSAGE_ID: 'store/buybutton.add-failure',
  SEE_CART_ID: 'store/buybutton.see-cart',
  TOAST_TIMEOUT: 3000,
}

const CSS_HANDLES = ['buyButtonContainer', 'buyButtonText']

const isTooltipNeeded = ({ showTooltipOnSkuNotSelected, skuSelector }: any) => {
  if (showTooltipOnSkuNotSelected && !skuSelector.areAllVariationsSelected) {
    return {
      showTooltip: true,
      labelId: CONSTANTS.SELECT_SKU_ERROR_ID,
    }
  }

  return {
    showTooltip: false,
  }
}

const skuItemToMinicartItem = (item: any) => {
  return {
    // Important for the mutation
    id: item.skuId,
    seller: item.seller,
    options: item.options,
    quantity: item.quantity,

    // Fields for optmistic cart
    sellingPrice: item.price,
    skuName: item.variant,
    detailUrl: item.detailUrl,
    imageUrl: item.imageUrl,
    name: item.name,
    listPrice: item.listPrice,
    assemblyOptions: item.assemblyOptions,
    sellingPriceWithAssemblies: item.sellingPriceWithAssemblies,

    // Fields for Analytics
    brand: item.brand,
    category: item.category,
    productRefId: item.productRefId,
  }
}

const useCallCartFinishIfPending = (
  orderFormContext: any,
  isAddingToCart: any,
  addToCartAndFinish: any
) => {
  const orderFormLoading = orderFormContext && orderFormContext.loading

  useEffect(() => {
    if (!orderFormLoading && isAddingToCart) {
      addToCartAndFinish()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [orderFormLoading])
}

type BuyButtonProps = {
  skuItems?: Array<{
    skuId: string
    quantity: number
    seller?: string | number
    name: string
    price: number
    variant?: string
    brand: string
    options?: Array<{
      id: string
      quantity: number
      assemblyId: string
      seller?: string
    }>
  }>
  children: React.ReactNode
  isOneClickBuy?: boolean
  shouldOpenMinicart?: boolean
  shouldAddToCart?: boolean
  large?: boolean
  intl: any
  available?: boolean
  showTooltipOnSkuNotSelected?: boolean
  showToast?: (...args: any[]) => any
  onAddStart?: (...args: any[]) => any
  onAddFinish?: (...args: any[]) => any
  addToCart: (...args: any[]) => any
  setMinicartOpen: (...args: any[]) => any
  orderFormContext?: any
  disabled?: boolean
  customToastURL?: string
  checkoutUrl?: string
}

/**
 * BuyButton Component.
 * Adds a list of sku items to the cart.
 */
export const BuyButton = ({
  intl,
  large,
  addToCart,
  skuItems,
  onAddStart,
  onAddFinish,
  setMinicartOpen,
  available = true,
  orderFormContext,
  isOneClickBuy = false,
  children,
  disabled: disabledProp,
  shouldAddToCart = true,
  shouldOpenMinicart = false,
  showTooltipOnSkuNotSelected = true,
  checkoutUrl,
  customToastURL = checkoutUrl,
}: BuyButtonProps) => {
  const handles = useCssHandles(CSS_HANDLES)
  const [isAddingToCart, setAddingToCart] = useState(false)
  const { showToast } = useContext(ToastContext)
  const {
    skuSelector = {
      areAllVariationsSelected: true,
    },
  } = useProduct()

  const dispatch = useProductDispatch()
  const { settings = {}, showInstallPrompt } = usePWA() || {}
  // @ts-expect-error ts-migrate(2339) FIXME: Property 'rootPath' does not exist on type 'Runtim... Remove this comment to see the full error message
  const { rootPath = '' } = useRuntime()
  const { promptOnCustomEvent } = settings
  const translateMessage = useCallback((id) => intl.formatMessage({ id }), [
    intl,
  ])

  const orderFormItems = path(['orderForm', 'items'], orderFormContext)
  const { utmParams, utmiParams } = useMarketingSessionParams()

  const resolveToastMessage = (success: any, isNewItem: any) => {
    if (!success) return translateMessage(CONSTANTS.ERROR_MESSAGE_ID)
    if (!isNewItem) return translateMessage(CONSTANTS.DUPLICATE_CART_ITEM_ID)

    const isOffline = window && window.navigator && !window.navigator.onLine
    const checkForOffline = !isOffline
      ? translateMessage(CONSTANTS.SUCCESS_MESSAGE_ID)
      : translateMessage(CONSTANTS.OFFLINE_BUY_MESSAGE_ID)

    return checkForOffline
  }

  const toastMessage = ({ success, isNewItem }: any) => {
    const message = resolveToastMessage(success, isNewItem)

    const action = success
      ? {
          label: translateMessage(CONSTANTS.SEE_CART_ID),
          href: rootPath + customToastURL,
        }
      : undefined

    showToast({ message, action })
  }

  const fullCheckoutUrl = rootPath + checkoutUrl

  const beforeAddToCart = (event: any) => {
    event.stopPropagation()
    event.preventDefault()

    setAddingToCart(true)
    onAddStart && onAddStart()
  }

  const addToCartAndFinish = async () => {
    let showToastMessage: any

    try {
      // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
      const minicartItems = skuItems.map(skuItemToMinicartItem)
      const localStateMutationResult = !isOneClickBuy
        ? await addToCart(minicartItems)
        : null

      const linkStateItems =
        localStateMutationResult && localStateMutationResult.data.addToCart

      const callOrderFormDirectly = !linkStateItems

      let success: any = null

      if (callOrderFormDirectly) {
        const variables = {
          orderFormId: orderFormContext.orderForm.orderFormId,
          // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
          items: skuItems.map((item) => ({
            id: item.skuId,
            seller: item.seller,
            options: item.options,
            quantity: item.quantity,
          })),
          ...(utmParams ? { utmParams } : {}),
          ...(utmiParams ? { utmiParams } : {}),
        }

        const mutationRes = await orderFormContext.addItem({ variables })
        const { items } = mutationRes.data.addItem

        // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
        success = skuItems.filter(
          (skuItem) =>
            !!items.find(
              ({ id, seller }: any) =>
                id === skuItem.skuId && seller === skuItem.seller
            )
        )
        await orderFormContext.refetch().catch(() => null)
      }

      const addedItem =
        (linkStateItems &&
          // @ts-expect-error ts-migrate(2532) FIXME: Object is possibly 'undefined'.
          skuItems.filter(
            (skuItem) =>
              !!linkStateItems.find(
                ({ id, seller }: any) =>
                  id === skuItem.skuId && seller === skuItem.seller
              )
          )) ||
        success

      const foundItem =
        addedItem.length &&
        orderFormItems &&
        // @ts-expect-error ts-migrate(2571) FIXME: Object is of type 'unknown'.
        orderFormItems.filter(
          (item: any) =>
            item.id === addedItem[0].skuId &&
            item.seller === addedItem[0].seller &&
            !item.canHaveAttachment
        ).length > 0

      success = addedItem

      showToastMessage = () =>
        toastMessage({
          success: success && success.length >= 1,
          isNewItem: !foundItem,
        })

      /* PWA */
      if (promptOnCustomEvent === 'addToCart' && showInstallPrompt) {
        showInstallPrompt()
      }

      shouldOpenMinicart && !isOneClickBuy && setMinicartOpen(true)
    } catch (err) {
      console.error(err)
      showToastMessage = () => toastMessage(false)
    }

    setTimeout(() => {
      setAddingToCart(false)
      showToastMessage()
      if (isOneClickBuy) {
        // eslint-disable-next-line no-restricted-globals
        location.assign(fullCheckoutUrl)
      }

      onAddFinish && onAddFinish()
    }, 500)
  }

  const handleAddToCart = async (event: any) => {
    beforeAddToCart(event)
    await addToCartAndFinish()
  }

  useCallCartFinishIfPending(
    orderFormContext,
    isAddingToCart,
    addToCartAndFinish
  )

  const handleClick = (e: any) => {
    if (dispatch) {
      dispatch({ type: 'SET_BUY_BUTTON_CLICKED', args: { clicked: true } })
    }

    if (skuSelector.areAllVariationsSelected && shouldAddToCart) {
      if (orderFormContext && orderFormContext.loading) {
        // Just call the before add to cart function and the useEffect hook will call the finish part when apropriate
        beforeAddToCart(e)
      } else {
        handleAddToCart(e)
      }
    }
  }

  if (!skuItems) {
    return <ContentLoader />
  }

  const disabled = disabledProp || !available
  const unavailableLabel = (
    <FormattedMessage id="store/buyButton-label-unavailable">
      {(message) => <span className={handles.buyButtonText}>{message}</span>}
    </FormattedMessage>
  )

  const ButtonWithLabel = (
    <Button
      block={large}
      disabled={disabled}
      onClick={handleClick}
      isLoading={isAddingToCart}
    >
      {available ? children : unavailableLabel}
    </Button>
  )

  const { showTooltip, labelId } = isTooltipNeeded({
    showTooltipOnSkuNotSelected,
    skuSelector,
  })

  const tooltipLabel = showTooltip && (
    <span className={handles.errorMessage}>
      <FormattedMessage id={labelId} />
    </span>
  )

  return !showTooltip ? (
    ButtonWithLabel
  ) : (
    <Tooltip trigger="click" label={tooltipLabel}>
      {ButtonWithLabel}
    </Tooltip>
  )
}

export default BuyButton
