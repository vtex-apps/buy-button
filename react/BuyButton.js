import BuyButton from './components/buy-button/src/index'

export default BuyButton
