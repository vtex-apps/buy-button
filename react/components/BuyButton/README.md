# BuyButton
BuyButton is a canonical component that any VTEX app can import.

To import it into your code: 
```js
import BuyButton from 'vtex.store-components/BuyButton'
```

## Usage
You can use it in your code like a React component with the jsx tag: `<BuyButton />`. 
```html
<BuyButton skuItems={[{ id: '1', quantity: 1, seller: 1 }]}> 
  Buy!
</BuyButton>
```

| Prop name          | Type       | Description                                                                 |
| ------------------ | ---------- | --------------------------------------------------------------------------- |
| `skuItems`         | `Array!`   | SKU Items to be added to the cart                                           |
| `message`          | `Node!`    | Component children that will be displayed inside of the button              |
| `quantity`         | `Number`   | Quantity of the product sku to be added to the cart                         |
| `seller`           | `Number`   | Which seller is being referenced by the button                              |
| `isOneClickBuy`    | `Boolean`  | Should redirect to the checkout page or not                                 |
| `intl`             | `Object`   | Internationalization property to format text                                |

#### SKU Items Props

| Prop name          | Type       | Description                                                                 |
| ------------------ | ---------- | --------------------------------------------------------------------------- |
| `skuId`            | `String!`  | Specification of which product will be added to the cart                    |
| `quantity`         | `Number!`  | Quantity of the product sku to be added to the cart                         |
| `seller`           | `Number!`  | Which seller is being referenced by the button                              |


