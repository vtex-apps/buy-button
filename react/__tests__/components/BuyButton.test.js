import React from 'react'
import { shallow } from 'enzyme'

import BuyButton from '../../BuyButton'

describe('<BuyButton />', () => {
  let wrapper

  const renderComponent = customProps => {
    const props = {
      ...customProps,
      showToast: () => {},
    }

    const component = shallow(<BuyButton {...props}> Test </BuyButton>)

    return component
  }

  it('should be rendered', () => {
    const wrapper = renderComponent()
    expect(wrapper).toBeDefined()
    expect(wrapper).toBeTruthy()
  })

  it('should match snapshot normal', () => {
    const wrapper = renderComponent({ available: true, skuItems: [] })
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot unavailable', () => {
    const wrapper = renderComponent({ available: false, skuItems: [] })
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot loading', () => {
    const wrapper = renderComponent({ available: false })
    expect(wrapper).toMatchSnapshot()
  })
})
