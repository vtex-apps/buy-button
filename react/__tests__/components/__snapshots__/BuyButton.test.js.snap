// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BuyButton /> should match snapshot loading 1`] = `
<Fragment>
  <ContentLoader />
</Fragment>
`;

exports[`<BuyButton /> should match snapshot normal 1`] = `
<Fragment>
  <Button
    disabled={false}
    isLoading={false}
    onClick={[Function]}
    primary={true}
  >
     Test 
  </Button>
</Fragment>
`;

exports[`<BuyButton /> should match snapshot unavailable 1`] = `
<Fragment>
  <Button
    disabled={true}
    isLoading={false}
    onClick={[Function]}
    primary={true}
  >
    <FormattedMessage
      id="buyButton-label-unavailable"
    />
  </Button>
</Fragment>
`;
